src/InvariantTheory/invariant_rings.jl

```diff
@@ -1,130 +1,166 @@
 
-   function InvariantRing{T1, T2}(G::MatrixGroup{T1, T2}) where { T1 <: RingElem, T2 <: MatrixElem{T1} }
-      IR = InvariantRing{T1, T2}([ g.elm for g in G.gens ])
-      IR.group = G
-      return IR
-   end
-end
+action(I::InvRing) = I.action
+
+_action_singular(I::InvRing) = I.action_singular
+
+group(I::InvRing) = I.group
 
-function InvariantRing(matrices::Vector{T2}) where T2 <: MatrixElem{fmpq}
-   @assert !isempty(matrices)
+ismodular(I::InvRing) = I.modular
 
-   return InvariantRing{elem_type(base_ring(matrices[1])), T2}(matrices)
+function invariant_ring(M::Vector{<: MatrixElem})
+  return invariant_ring(base_ring(M[1]), M)
 end
 
-function InvariantRing(G::MatrixGroup{T1, T2}) where { T1 <: fmpq, T2 <: MatrixElem{T1} }
-   return InvariantRing{T1, T2}(G)
+function invariant_ring(K::Field, M::Vector{<: MatrixElem})
+  return invariant_ring(matrix_group([change_base_ring(K, g) for g in M]))
 end
 
-InvariantRing(matrices::MatrixElem{fmpq}...) = InvariantRing(collect(matrices))
+function invariant_ring(G::MatrixGroup)
+  n = degree(G)
+  R, = PolynomialRing(base_ring(G), "x" => 1:n, cached = false)
+  S = singular_ring(R)
+  action = mat_elem_type(typeof(G))[g.elm for g in gens(G)]
+  return InvRing(base_ring(G), R, G, action)
+end
 
-function Base.show(io::IO, IR::InvariantRing)
-   if isdefined(IR, :group)
-      print(io, "Invariant ring of group ")
-      print(io, IR.group)
-   else
-      print(io, "Invariant ring of matrix group generated by ")
-      print(io, IR.matrices)
-   end
+invariant_ring(matrices::MatrixElem{T}...) where {T} = invariant_ring(collect(matrices))
+
+function Base.show(io::IO, IR::InvRing)
+  print(io, "Invariant ring of\n")
+  print(io, group(IR), "\n")
+  print(io, "with generators\n")
+  print(io, action(IR))
 end
 
-function reynolds_molien_via_singular(IR::InvariantRing)
+function reynolds_molien_via_singular(IR::InvRing)
    if !isdefined(IR, :reynolds_singular) || !isdefined(IR, :molien_singular)
-      R = IR.poly_ring
-      RS = Oscar.singular_ring(R)
-      singular_matrices = Vector{Singular.smatrix{elem_type(RS)}}(undef, length(IR.matrices))
-      for i = 1:length(IR.matrices)
-         singular_matrices[i] = map_entries(RS, IR.matrices[i])
-      end
-
+      R = polynomial_ring(IR)
+      singular_matrices = _action_singular(IR)
       rey, mol = Singular.LibFinvar.reynolds_molien(singular_matrices...)
       IR.reynolds_singular = rey
       IR.molien_singular = mol
    end
    return IR.reynolds_singular, IR.molien_singular
 end
 
-function primary_invariants_via_singular(IR::InvariantRing)
-   @assert characteristic(IR.field) == 0
+function primary_invariants_via_singular(IR::InvRing)
    if !isdefined(IR, :primary_singular)
-      rey, mol = reynolds_molien_via_singular(IR)
-      P = Singular.LibFinvar.primary_char0(rey, mol)
-      p = Vector{elem_type(IR.poly_ring)}()
-      for i = 1:ncols(P)
-         push!(p, IR.poly_ring(P[1, i]))
+      if ismodular(IR)
```

Comment:
Here it should be `!ismodular(IR)`, no? But actually one can so far just always call primary_invariants in LibFinvar. I just started doing the reynolds operator and such separately to be able to replace single parts at some point.

Reply:
I wanted to try an example with a finite field, but they are bugged anyway (see the recent Oscar issues).